language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -197,48 +202,99 @@ data CostModel =
     , paramGtByteString         :: CostingFunTwoArguments
     , paramIfThenElse           :: CostingFunThreeArguments
     }
-    deriving (Show, Eq, Generic, Lift)
+    deriving (Show, Eq, Generic, Lift, Default, NFData)
     deriving (FromJSON, ToJSON) via CustomJSON '[FieldLabelModifier (StripPrefix "param", CamelToSnake)] CostModel
 
 -- TODO there's probably a nice way to abstract over the number of arguments here. Feel free to implement it.
 -- TODO loglinear is currently linear
 
-data CostingFunOneArgument =
-    OneArgumentConstantCost Integer
-    | OneArgumentLogLinearCost Integer
-    deriving (Show, Eq, Generic, Lift)
-    deriving (FromJSON, ToJSON) via CustomJSON '[SumTaggedObject "type" "arguments", ConstructorTagModifier (StripPrefix "OneArgument", CamelToSnake)] CostingFunOneArgument
+data CostingFunOneArgument = CostingFunOneArgument
+    { costingFunOneArgumentCpu    :: ModelOneArgument
+    , costingFunOneArgumentMemory :: ModelOneArgument
+    }
+    deriving (Show, Eq, Generic, Lift, Default, NFData)
+    deriving (FromJSON, ToJSON) via CustomJSON
+        '[FieldLabelModifier (StripPrefix "costingFunOneArgument", CamelToSnake)] CostingFunOneArgument
+
+data ModelOneArgument =
+    ModelOneArgumentConstantCost Integer
+    | ModelOneArgumentAddedSizes ModelAddedSizes
+    deriving (Show, Eq, Generic, Lift, NFData)
+    deriving (FromJSON, ToJSON) via CustomJSON
+        '[SumTaggedObject "type" "arguments", ConstructorTagModifier (StripPrefix "ModelOneArgument", CamelToSnake)] ModelOneArgument
+instance Default ModelOneArgument where
+    def = ModelOneArgumentConstantCost 1
 
 runCostingFunOneArgument :: CostingFunOneArgument -> ExMemory -> ExBudget
 runCostingFunOneArgument
-    (OneArgumentConstantCost c) _ = ExBudget (ExCPU c) (ExMemory c)
-runCostingFunOneArgument
-    (OneArgumentLogLinearCost factor) (ExMemory size1) = ExBudget (ExCPU $ size1 * factor) (ExMemory $ size1 * factor)
+    (CostingFunOneArgument cpu mem) mem1 =
+        ExBudget (ExCPU (runOneArgumentModel cpu mem1)) (ExMemory (runOneArgumentModel mem mem1))
+
+runOneArgumentModel :: ModelOneArgument -> ExMemory -> Integer
```

Comment:
Have we got a note somewhere about the general approach with these `Model` types?

Reply:
Yeap, see `CostModel.md`.